src/cmap/connection.ts

```diff
@@ -444,34 +460,13 @@ export class Connection extends TypedEventEmitter<ConnectionEvents> {
   }
 
   destroy(options: DestroyOptions, callback?: Callback): void {
-    this.removeAllListeners(Connection.PINNED);
-    this.removeAllListeners(Connection.UNPINNED);
-
-    if (this[kStream] == null || this.destroyed) {
-      this.destroyed = true;
-      if (typeof callback === 'function') {
-        callback();
-      }
-
-      return;
-    }
-
-    if (options.force) {
-      this[kStream].destroy();
-      this.destroyed = true;
-      if (typeof callback === 'function') {
-        callback();
-      }
-
+    if (this.closed) {
+      process.nextTick(() => callback?.());
       return;
     }
-
-    this[kStream].end(() => {
-      this.destroyed = true;
-      if (typeof callback === 'function') {
-        callback();
-      }
-    });
+    this.once('close', () => process.nextTick(() => callback?.()));
```

Comment:
I am happy to revert this change. By listening for the close event, the centralized cleanup handler does not need to worry about an optional callback. None of the event handlers rely on it, so it's only necessary in this method.

Reply:
I was thinking we could make cleanup an async function and we can invoke the callback from a `.then`. If that's reasonable let me know. Otherwise, we can keep this but add a gate: `if (callback) this.once ///...`